src/ros_ibeo_lux.cpp

```diff
@@ -181,145 +183,135 @@ int main(int argc, char **argv)
-
               if (first_mw > 0)
                 grand_buffer.erase(
                     grand_buffer.begin(),
                     grand_buffer.begin() + first_mw);  // Unusable data in beginning of buffer.
 
               // From here on, the detected Magic Word should be at the beginning of the grand_buffer.
 
-              // ROS_DEBUG("Size before reading message: %lu", grand_buffer.size());
-
               IbeoDataHeader header;
-              std::vector<unsigned char> msg;
-
-              header.parse(grand_buffer.data());
+              header.parse(std::vector<uint8_t>(grand_buffer.begin(), grand_buffer.begin() + IBEO_HEADER_SIZE));
 
-              // ROS_DEBUG("Calculated size of message: %u", IBEO_HEADER_SIZE + header.message_size);
+              auto total_msg_size = IBEO_HEADER_SIZE + header.message_size;
 
-              if (grand_buffer.size() < IBEO_HEADER_SIZE + header.message_size)
+              if (grand_buffer.size() < total_msg_size)
                 break;  // Incomplete message left in grand buffer. Wait for next read.
 
-              msg.insert(
-                  msg.end(),
+              std::vector<uint8_t> msg(
                   grand_buffer.begin(),
-                  grand_buffer.begin() + IBEO_HEADER_SIZE + header.message_size);
-              // ROS_DEBUG("Size of copied message: %lu", msg.size());
-              messages.push_back(msg);
+                  grand_buffer.begin() + total_msg_size);
+              messages.push(msg);
               grand_buffer.erase(
                   grand_buffer.begin(),
-                  grand_buffer.begin() + IBEO_HEADER_SIZE + header.message_size);
-
-              // ROS_DEBUG("Size after reading message: %lu", grand_buffer.size());
+                  grand_buffer.begin() + total_msg_size);
             }
 
             if (!ros::ok())
               break;
           }
         }
 
-        if (!messages.empty())
+        while (!messages.empty())
         {
-          for (unsigned int i = 0; i < messages.size(); i++)
+          auto message = messages.front();
+
+          network_interface::TCPFrame raw_frame;
+          raw_frame.address = ip_address;
+          raw_frame.port = port;
+          raw_frame.size = message.size();
+          raw_frame.data.insert(raw_frame.data.end(), message.begin(), message.end());
+          raw_frame.header.frame_id = frame_id;
+          raw_frame.header.stamp = ros::Time::now();
+
+          raw_tcp_pub.publish(raw_frame);
+
+          IbeoDataHeader ibeo_header;
+          ibeo_header.parse(message);
+
+          // Instantiate a parser class of the correct type.
+          auto class_parser = IbeoTxMessage::make_message(ibeo_header.data_type_id);
+          // Look up the message publisher for this type.
+          auto pub = pub_list.find(ibeo_header.data_type_id);
+
+          // Only parse message types we know how to handle.
+          if (class_parser != NULL && pub != pub_list.end())
```

Comment:
The while loop on ln243 is entered because messages has some content. But the only place where that content seems to be removed is within the ln266 condition statement. If this condition fails, it appears the program will be stuck in an endless loop.

Reply:
You're right. This was a typo. It should have been out another level. Fixed.